src/main/java/org/breedinginsight/brapps/importer/daos/BrAPIObservationUnitDAO.java

```diff
@@ -114,8 +123,16 @@ public List<BrAPIObservationUnit> getObservationUnitsForTrialDbId(@NotNull UUID
         observationUnitSearchRequest.trialDbIds(List.of(trialDbId));
 
         ObservationUnitsApi api = brAPIEndpointProvider.get(programDAO.getCoreClient(program.getId()), ObservationUnitsApi.class);
-        return brAPIDAOUtil.search(api::searchObservationunitsPost,
+        List<BrAPIObservationUnit> observationUnits = brAPIDAOUtil.search(api::searchObservationunitsPost,
                 api::searchObservationunitsSearchResultsDbIdGet,
                 observationUnitSearchRequest);
+        if( withGID ){
+            for (BrAPIObservationUnit observationUnit: observationUnits) {
+                String germplasmDbId = observationUnit.getGermplasmDbId();
+                BrAPIGermplasm germplasm = this.germplasmService.getGermplasmByDBID(program.getId(), germplasmDbId).get() ;
+                observationUnit.putAdditionalInfoItem("gid", germplasm.getAccessionNumber());
+            }
```

Comment:
```suggestion
            // Load germplasm for program into map.
            // TODO: if we use redis search, that may be more efficient than loading all germplasm for the program.
            HashMap<String, BrAPIGermplasm> germplasmByDbId = new HashMap<>();
            this.germplasmService.getGermplasm(programId).forEach((germplasm -> germplasmByDbId.put(germplasm.getGermplasmDbId(), germplasm)));
            for (BrAPIObservationUnit observationUnit: observationUnits) {
                BrAPIGermplasm germplasm = germplasmByDbId.get(observationUnit.getGermplasmDbId());
                observationUnit.putAdditionalInfoItem("gid", germplasm.getAccessionNumber());
            }
```
This is loading all Germplasm in the program from the cache for each observationUnit, which really hurts performance for large datasets. I would load all germplasm for the program into a HashMap (key: germplasmDbId => value: Germplasm object) outside of the for loop and then access by key inside the loop.

I think there's an eventual desire to use something like https://redis.io/docs/interact/search-and-query/ to make more fine-grained cache retrievals, but for now, just use a HashMap.

Reply:
Change looks good to me

Reply:
Excellent!  Thank you.
Done.